### DISCUSSION TITLE
Create Common Managed Base Dialog Library

### DISCUSSION THREAD

**Discussion by bhagatus** [oleg-shilo/wixsharp#1698] [General]
I have multiple setups and need to create dialogs with same background theme and common functionalities across all the setups.
I was planning to create a library of custom base dialog inherited from ManagedForm and IManagedDialog. 
I wanted to use this library to be utilized by all setups in different setup projects. 

But I am failing to create such library for Managed Custom Dialog. 
I can see it can be done with External Custom UI but I wanted to utilize the functionality of Managed Form.

Do you have any suggestion for any of approach?

**Comment by oleg-shilo** [comment]
I would expect that you can start with the standard template project, then you can replace the `ManagedForm` with its derived class:

![image](https://github.com/user-attachments/assets/221e4f73-7f7b-4395-89e3-654b55db42f8)

```c#
    public partial class CustomDialog : ManagedForm, IManagedDialog
    {
```

**Comment by bhagatus** [comment]
Hi Oleg,

I have created sample application to utilize common dialog.
First example is using WIxSharp dialog template  (CommonDialog and CommonDialogWix_Consumer)
And other example is using Windows Form project and referencing WixSharp.bin to inherit the managed form. (CommonDialogWin and CommonDialogWin_Consumer)

In both samples I have inherited the CommonDialog in WelcomeDialog only for testing.

In both cases, I am able to built the MSI. But MSI is getting error as shown in log.log file created in each folder

Summary of error is 
[Uploading WixSharp4_CommonDialog_Sample.zip…]()




	Cannot attach ManagedUI: System.Exception: Cannot instantiate 'MainSetup_WixCommonDialogConsumer.Dialogs.WelcomeDialog'. Make sure you added this type of assembly to your setup with 'project.DefaultRefAssemblies'

**Reply by oleg-shilo** [reply]
The sample link is not working.

**Comment by bhagatus** [comment]
[WixSharp4_CommonDialog_Sample.zip](https://github.com/user-attachments/files/18183472/WixSharp4_CommonDialog_Sample.zip)

This link should work.

**Comment by oleg-shilo** [comment]
The problem is caused by the fact that the assembly implementing your CustomDialog is not included in the setup so it's not available at runtime on the target system and your derived class (Form) cannot be instantiated.

You can easily correct it with this line:
```c#
project.DefaultRefAssemblies.Add(typeof(CustomDialog).Assembly.Location);
```

**Comment by bhagatus** [comment]
Hello,

Thanks for the input. I am able to inherit the base dialog created in separate project. It is working in both options by referencing project directly and referencing Nuget package.
The only thing I need to check that instead of using WixSharp template for base dialog , can I use WIxSharp bin only. I don't want to create MSI for BaseDialog. I will check it. 

But my basic problem is resolved. Thanks for your input.
